### ISSUE TITLE
Retrieve phase envelope through high-level DLL

### ISSUE THREAD

**Issue by leolima75** [CoolProp/CoolProp#1158]
Hi,

Great work with Coolprop ! I have been using the C# wrapper just fine, but then I'd like to call Coolprop from within Ironpython scripts ... Since I couldn't find a way to install Coolprop pyhton package (is it possible anyway ?), I was looking for ways to call the C Std dll functions directly. I found the code below

```
CP = ctypes.cdll.LoadLibrary("CoolProp_stdC.dll") # renamed the coolprop library to reflect the version
PropsSI = CP.PropsSI
PropsSI.argtypes = (ctypes.c_char_p, ctypes.c_char_p, ctypes.c_double, ctypes.c_char_p, ctypes.c_double, ctypes.c_char_p)
PropsSI.restype = ctypes.c_double
```

Which works just fine. But I'd like to get the phase envelope using the AbstractState_factory and couldn't find code for it (nor could I write code that works - I just keep crashing Ironpython ...)

Could you help with that ?  Thanks

**Comment by ibell** [comment]
There's this link: http://stackoverflow.com/a/34630229  Getting python working natively in IronPython is obviously your best option.

The phase envelope is not exposed in the DLL.  If you need that, feel free to make the modifications in the CoolPropLib.[h/cpp] files.  @JonWel might be able to help.

**Comment by ibell** [comment]
Any updates?  Sounds like IronPython is not currently being updated very much....

**Comment by leolima75** [comment]
Hello Dr. Ian,

    Thank you for the interest ! Unfortunately I'm not much of a programmer to even think about trying to dig inside Coolprop code ... Yes, it looks like Ironpython hasn't been upodate for a while, but still it's the best way I know for integrating python scripts with Visual Studio C# interfaces.

    The solution I talked about in my first comment 'half' solved my problem. What I been trying to acomplish is to setup an interface to evaluate the adiabatic expansion of water saturated natural gas (NG) through compressors' dry gas seals. This requires three steps (involving thermodynamic calculations), calculate the water amount that saturates the gas at a given P and T (since we only have de HC/CO2/N2 data from field analysers), calculate de phase envelope for the saturated mixture and finally calculate the adiabatic expansion from compressors' discharge conditions to nearly atmospheric pressure. For some NG mixes it's working good (at least it looks like), bute for others I can't seem to get the water saturation routine to converge and, also, there are cases where the phase envelope routine (using PropsSI) won't work for the saturated gas (calculated).

![dry_gas](https://cloud.githubusercontent.com/assets/4535785/17586944/585fc450-5f9b-11e6-97e1-431278257035.png)

**Comment by ibell** [comment]
OK, nice to see you are now at least able to call the DLL using ctypes.  If you want to use mixtures through the DLL, you might have to use the low-level interface through the high-level interface (http://www.coolprop.org/dev/coolprop/LowLevelAPI.html#access-from-high-level-interface).  It is far less convenient, but might be needed because our support for mixtures through PropsSI is quite limited.

I'm planning on some updates to mixture flash calculations soon because I need them for work.  Watch this space.

**Comment by leolima75** [comment]
Thanks, I really appreciate it ! I'll look into it and keep an eye for future Coolprop developments.

**Comment by ibell** [comment]
I guess we are going to close this issue for now.  Thanks for the info.

**Comment by ibell** [comment]
I just re-read the original question.  You want to retrieve the phase envelope through the DLL, which currently is not possible :(

**Comment by leolima75** [comment]
Ok, I've have been trying to call the low-level interface through ctypes, but haven't been succesful so far. I can't see whether there's something wrong with the inputs ... And keyed_output is returning an int which doesn't make sense to me. Part of the code is as follows:

CP = ctypes.cdll.LoadLibrary("CoolProp_cdecl.dll")

feedback = ''

indPair  = CP.get_input_pair_index(ctypes.c_char_p('PQ_INPUTS')) 
feedback += 'pair index ' + str(indPair) + '\n\n'
indParam  = CP.get_param_index(ctypes.c_char_p('T')) 
feedback += 'param index ' + str(indParam) + '\n\n'

for m in range(0,len(x)):  feedback += gas[m] + '  ' + str(x[m]) + '\n'

feedback += '\n'

fluido = ''
for m in range(len(gas)):
  if (m < len(gas)-1) :
    fluido += gas[m] + '&'
  else : 
    fluido += gas[m]
fluido += '&WATER'

feedback += 'gas ' + fluido + '\n\n'

ierr= ctypes.c_long(0)
herr= ctypes.create_string_buffer('', 500)
handle = CP.AbstractState_factory(ctypes.c_char_p('HEOS'),ctypes.c_char_p(fluido),ctypes.byref(ierr),ctypes.byref(herr),ctypes.c_long(500))
feedback += 'status de erro ' + str(ierr.value) + ' texto do erro ' + str(herr.value) + '\n\n'

TDew = 30.5 + 273.15
DComp = 0.002
x1 = [x[0]]
for m in range(1,len(x)): append(x1, [(1 - DComp) * x[m]])
append(x1, [DComp])

XComp = (ctypes.c_double*(len(x)+1))(*x1)

CP.AbstractState_set_fractions(handle,XComp,ctypes.c_long(len(x)+1),ctypes.byref(ierr),ctypes.byref(herr),ctypes.c_long(500))
feedback += 'status de erro ' + str(ierr.value) + ' texto do erro ' + str(herr.value) + '\n\n'

CP.AbstractState_update(handle,indPair,ctypes.c_double(Pi),ctypes.c_double(1),ctypes.byref(ierr),ctypes.byref(herr),ctypes.c_long(500))
feedback += 'Pi ' + str(Pi) + '\n\n'
feedback += 'status de erro ' + str(ierr.value) + ' texto do erro ' + str(herr.value) + '\n\n'

TComp = CP.AbstractState_keyed_output(handle,indParam,ctypes.byref(ierr),ctypes.byref(herr),ctypes.c_long(500))

feedback += 'status de erro ' + str(ierr.value) + ' texto do erro ' + str(herr.value) + '\n\n'

feedback += 'TDew = ' + str(TComp) + '\n\n'

Which gives back:

pair index 2

param index 19

NITROGEN  0.00914
CO2  0.00926
METHANE  0.85047
ETHANE  0.0848
PROPANE  0.03548
ISOBUTAN  0.0042
BUTANE  0.00573
IPENTANE  0.00051
PENTANE  0.00041

gas NITROGEN&CO2&METHANE&ETHANE&PROPANE&ISOBUTAN&BUTANE&IPENTANE&PENTANE&WATER

status de erro 0 texto do erro 

status de erro 0 texto do erro 

Pi 1678234.32463

status de erro 1 texto do erro Error: Residual function in secant returned invalid number

status de erro 0 texto do erro Error: Residual function in secant returned invalid number

TDew = 532915788

**Comment by leolima75** [comment]
Hello Dr. Bell,
    Thank you very much ! Really apreciate it. Right now I'm doing shipyard work, but I'll give it a try as soon as I'm back in the office.

Enviado do Yahoo Mail no Android 
 
  Em 22:35 Qui, 13 de out de PM, Ian Bell<notifications@github.com> escreveu:   
Closed #1158 via c929e92.

**Comment by ibell** [comment]
Let me know how you get on...
